Update What's New for .NET 10 with P3 info (#45765)
dotnet/docs: docs/core/whats-new/dotnet-10/libraries.md

Reviewer: Can you also remove the square brackets on line 136 and add in xrefs for APIs that were introduced in P1/P2?

        IServiceProvider? serviceProvider = null, 
        IDictionary<object, object?>? items = null);
}
```

## Support for telemetry schema URLs in `ActivitySource` and `Meter`

<xref:System.Diagnostics.ActivitySource> and <xref:System.Diagnostics.Metrics.Meter> now support specifying a telemetry schema URL during construction, aligning with OpenTelemetry specifications. This ensures consistency and compatibility for tracing and metrics data. Additionally, the update introduces `ActivitySourceOptions`, simplifying the creation of <xref:System.Diagnostics.ActivitySource> instances with multiple configuration options.

```csharp
public sealed partial class ActivitySource
{
    public ActivitySource(ActivitySourceOptions options);
    public string? TelemetrySchemaUrl { get; }
}

public class ActivitySourceOptions
{
    public ActivitySourceOptions(string name);
    public string Name { get; set; }
    public string? Version { get; set; }
    public IEnumerable<KeyValuePair<string, object?>>? Tags { get; set; }
    public string? TelemetrySchemaUrl { get; set; }
}

public partial class Meter : IDisposable
{
    public string? TelemetrySchemaUrl { get; }
}
```

> 

Reviewer: Since the ML.NET updates aren't versioned with .NET 10, I'm not sure if it makes sense to put this and the next update in this article. For .NET 9, we decided to maintain a separate [What's new in ML.NET article](https://learn.microsoft.com/en-us/dotnet/machine-learning/whats-new/overview). If that article is no longer being updated, then we should probably delete it and combine the info into the existing What's new in .NET 9/10 articles.

Author: It looks like it hasn't been maintained since last year.

Author: Resolving this for now because I need to get this out the door today. We can figure out what to do with ML.NET before Preview 4 ships

> ## Byte-Level support in BPE tokenizer

Reviewer: ```suggestion
## Byte-level support in BPE tokenizer
```

> The <xref:Microsoft.ML.Tokenizers.BpeTokenizer> now supports Byte-Level encoding, enabling compatibility with models like DeepSeek. This enhancement processes vocabulary as UTF-8 bytes, simplifying tokenizer creation with the new `BpeOptions` type.

Reviewer: ```suggestion
The <xref:Microsoft.ML.Tokenizers.BpeTokenizer> now supports byte-level encoding, enabling compatibility with models like DeepSeek. This enhancement processes vocabulary as UTF-8 bytes. In addition, the new `BpeOptions` type simplifies tokenizer configuration.
```

```csharp
BpeOptions bpeOptions = new BpeOptions(vocabs);
BpeTokenizer tokenizer = BpeTokenizer.Create(bpeOptions);
```

## Deterministic option for LightGBM trainer in ML.NET

LightGBM trainers now expose options for deterministic training, ensuring consistent results with the same data and random seed. These options include `deterministic`, `force_row_wise`, and `force_col_wise`.

```csharp
LightGbmBinaryTrainer trainer = ML.BinaryClassification.Trainers.LightGbm(new LightGbmBinaryTrainer.Options
{
    Deterministic = true,
    ForceRowWise = true
});
```

## Tensor enhancements

> The <xref:System.Numerics.Tensors.Tensor> class now includes a non-generic interface for operations like accessing `Lengths` and `Strides`. Slice operations no longer copy data, improving performance. Additionally, data can be accessed non-generically by boxing to `object` when performance is not critical.

Reviewer: ```suggestion
The <xref:System.Numerics.Tensors.Tensor> class now includes a nongeneric interface for operations like accessing `Lengths` and `Strides`. Slice operations no longer copy data, improving performance. Additionally, data can be accessed nongenerically by boxing to `object` when performance isn't critical.
```
